ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -198,3 +233,309 @@ module Json = struct
   let api_errors =
     List.map (fun error -> `O [("name", `String error)]) !Api_errors.errors
 end
+
+module Types = struct
+  let rec decompose = function
+    | Set x as y ->
+        y :: decompose x
+    | Map (a, b) as y ->
+        (y :: decompose a) @ decompose b
+    | Option x as y ->
+        y :: decompose x
+    | Record r as y ->
+        let name = snake_to_camel r ^ "Record" in
+        let types_in_field =
+          List.assoc_opt name records
+          |> Option.value ~default:[]
+          |> List.concat_map (fun field -> decompose field.ty)
+        in
+        y :: types_in_field
+    | (SecretString | String | Int | Float | DateTime | Enum _ | Bool | Ref _)
+      as x ->
+        [x]
+
+  let mesages objects = objects |> List.concat_map (fun x -> x.messages)
+
+  (** All types of params in a list of objects (automatically decomposes) *)
+  let of_params objects =
+    let param_types =
+      mesages objects
+      |> List.concat_map (fun x -> x.msg_params)
+      |> List.map (fun p -> p.param_type)
+      |> Listext.List.setify
+    in
+    List.concat_map decompose param_types |> Listext.List.setify
+
+  (** All types of results in a list of objects (automatically decomposes) *)
+  let of_results objects =
+    let return_types =
+      let aux accu msg =
+        match msg.msg_result with None -> accu | Some (ty, _) -> ty :: accu
+      in
+      mesages objects |> List.fold_left aux [] |> Listext.List.setify
+    in
+    List.concat_map decompose return_types |> Listext.List.setify
+end
+
+module Convert = struct
```

Comment:
I'm happy with this module now unless other reviewers have other thoughts.
Minor comment on the "func_partial_type"/"fp_type". The naming looks not great. This was also raised in another PR.